### PR TITLE
N°5794 - Add support for placeholders (recipient and quiz_url) in subject and body

### PR DESCRIPTION
**Pull request by jbostoen**
First of all, kudos on offering this extension!

It's exactly what I wanted to do from within iTop.

Except for one thing...

Right now, I can configure a proper subject and body for the e-mail. 
However, the "Click here to start the quizz" bothered me a bit.

I mean, my use case is that I want to write a nice greeting ("Hi xxx") rather than a generic one ("Hi customer", "Hi friend", "Hi").

This pull request allows to use some placeholders:
* `$recipient->attcode$` (contact information)
* `$on_behalf_of->attcode$` (contact information of the "on behalf of" contact)
* `$quiz_url$` (link to the quiz, with token. Same as the one for the default message).
* It also exposes the default placeholders such as `$current_contact->attcode$` (more of a side effect I guess).

Now, when there is no `$quiz_url$` in the email body, it defaults to the previous behavior and does add the default sentence/link at the bottom.

PS: it may also be worth to open a separate discussion about the spelling. Shouldn't this be "quiz" (in English) instead of "quizz"?

PSS: also in the datamodel, what is `<jointype/>`?

### CODE REVIEW

File: datamodel.customer-survey.xml
Comment (Molkobain) [inline datamodel.customer-survey.xml:1896]:
```suggestion
		if(preg_match('/\$quizz_url\$/', $sBody)) {
```

File: datamodel.customer-survey.xml
Comment (Molkobain) [inline datamodel.customer-survey.xml:1887]:
```suggestion
		$aParams['quizz_url'] = $sQuizzUrl;
```

**Review by Molkobain — APPROVED** [review]

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1892]: Would be more flexible if we could just use `$this$` and all the options from that instead of only the `on_behalf_of` field.

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1898]:
```suggestion
		// Email body contains a placeholder?
		$bPrependLink = !preg_match('/\$quiz_url\$/', $sBody);
```

File: datamodel.customer-survey.xml
Comment (jbostoen) [reply datamodel.customer-survey.xml:1892]: Hm, I'm not entirely sure, can I just refer to "this"? (I know there's a function to turn an object into query arguments, just not sure if it would also work out of the box to support `$this->on_behalf_of->attcode$`

Comment (Hipska) [reply datamodel.customer-survey.xml:1892]: that's the whole purpose of it indeed..

**Comment by Molkobain** [comment]
`on_behalf_of` can be removed from the description I guess

**Comment by Molkobain** [comment]
Good enhancement 🙌

**Comment by jbostoen** [comment]
> Good enhancement 🙌

I hope so, seems like quite a generic use case :D

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1897]: Actually, the link is being appended, not prepended. The variable name can cause confusion..

Comment (jbostoen) [reply datamodel.customer-survey.xml]: You are absolutely right.

**Comment by Molkobain** [comment]
The idea was accepted during technical review. But we would like you to add a unit test to ensure the placeholders keep working in the future. Send a message to @piRGoif for any help :)

_PR will be reviewed functionally next week anyway to save time for the next iteration._

**Comment by Molkobain** [comment]
Accepted during functional review.
Pending for unit test before merge.

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: The suggestion is included in the current code. Closing conversation.

Comment (Hipska) [reply datamodel.customer-survey.xml]: Indeed, I couldn't mark as resolved. I think only author and maintainers can?

Comment (piRGoif) [reply datamodel.customer-survey.xml]: I think so !

**Comment by piRGoif** [comment]
Hello @jbostoen , I've just pushed 2 commits with a new test class : test/SurveyTest.php
The first commit is some code refactoring to ease the test.

You can try to launch the test class on your env ?
You'll see the \Combodo\iTop\CustomerSurvey\Test\SurveyTest::ApplyParamsToContentProvider method contains only one test case, you should add more.

File: datamodel.customer-survey.xml
Comment (jbostoen) [inline datamodel.customer-survey.xml:1933]: Mind that in English it's actually a single Z :)

File: datamodel.customer-survey.xml
Comment (jbostoen) [inline datamodel.customer-survey.xml:1895]: I honestly can imagine very few scenarios where someone wants to put the URL in the subject; but why would we make it impossible?

File: datamodel.customer-survey.xml
Comment (jbostoen) [inline datamodel.customer-survey.xml:1973]: I actually shorted this based on @Hipska 's suggestion; why should we make it longer again?

File: datamodel.customer-survey.xml
Comment (Hipska) [reply datamodel.customer-survey.xml:1895]: I was thinking this as well, but there is a small detail, the URL parameter is set to `null` here.

Comment (piRGoif) [reply datamodel.customer-survey.xml:1895]: The intent is not very clear here indeed : I passed null mainly because we never want to append the link on the subject.
So maybe we should add a boolean parameter for this behavior ? Something like `bShouldAppendLinkIfNoUrlPlaceholder` ?

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Woops, thanks !
Fixed in be41bc6f

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: I made those modifications for 2 reasons : 
- I prefer having an explicit variable name than a variable name + an inline comment
- preg_match doesn't return a boolean (1, 0 or false), so I prefer using a true boolean in the test

**Comment by piRGoif** [comment]
Sorry I'll be on holidays for a week and a half... But other Combodo dev should be able to follow up

File: datamodel.customer-survey.xml
Comment (Hipska) [reply datamodel.customer-survey.xml]: I agree the variable name, the rest can be simplified as:

```php
     if ($quizUrlPatternsFound === 1) {
         return $sContent;
     }

     return $sContent . '<br/><a href="'.$sQuizUrl.'">'.Dict::S('Survey-notif-linktoquizz').'</a>';
```

File: datamodel.customer-survey.xml
Comment (Hipska) [reply datamodel.customer-survey.xml:1895]: It's a bit long, no? `bAutoAppendLink` should be enough, you still have the php doc that can explain more if needed.

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Thanks, really like the idea ! 
Pushed in 1328f76d

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1946]: This if seems not needed? Or will the ApplyParams fail in that case?

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1951]: I think the ordering of these calls needs to be reversed?

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1971]: It doesn't make sense to call this method if there is no URL.

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1974]: Could maybe be even more simplified, or would that be too dense according to you?
```suggestion
    if (preg_match('/\$quiz_url\$/', $sContent)=== 1) {
```

It makes sense to me as the var isn't used anymore after that..

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Done in 770f2437

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Was a enormous bug, I wrote very buggy code, my apologies.
Fixed in d31a92b2

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Fixed !

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Fixed !

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml:1974]: Fixed !

**Comment by piRGoif** [comment]
Ok I made a lot of modifications in the latest commits... Now I'm really heading for holidays :o)
Merry Christmas & happy new year !

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1971]: Again, it doesn't make sense to call this method if not wanted to append a link..

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1948]:
```suggestion
  if($bAutoAppendLink) $sContent = $this->AppendLinkToContent($sContent, $sQuizUrl);
```

File: datamodel.customer-survey.xml
Comment (Hipska) [reply datamodel.customer-survey.xml:1974]: .. but not pushed?

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1933]:
```suggestion
	 * @param string $sQuizUrl
```

**Review by Hipska — CHANGES_REQUESTED** [review]

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml:1971]: Right, I agree.
There are some adaptations to do... @jbostoen don't hesitate to change it.
Otherwise it will wait after I'm back from holidays :/

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml:1974]: Oh sorry I misunderstood, I though this was about the 2 variables that were used before.
I like the intermediate variable : that explains what the pattern means.

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Should be fixed in 45c45820 !

File: datamodel.customer-survey.xml
Comment (piRGoif) [reply datamodel.customer-survey.xml]: Should be fixed in 45c45820 !

**Comment by piRGoif** [comment]
Hello,
And happy new year 🥳

I've just pushed a fix for the last review comments.
Can you check if everything is ok ?

**Comment by piRGoif** [comment]
Hello,
The next review is planned on next tuesday. @jbostoen are you ok with my changes ?

File: datamodel.customer-survey.xml
Comment (Hipska) [inline datamodel.customer-survey.xml:1964]: Not really important, but now these comments will look weird in the generated php code (way too much indentation)

**Review by Hipska — APPROVED** [review]

**Comment by jbostoen** [comment]
> Hello, The next review is planned on next tuesday. @jbostoen are you ok with my changes ?

Sure :)

**Review by Molkobain — APPROVED** [review]

**Comment by piRGoif** [comment]
Thanks for the answer Jeffrey ! I hope the PR could be reviewed tomorrow.

**Comment by Molkobain** [comment]
> Accepted during functional review. Pending for unit test before merge.

I forgot that PR was already accepted during both reviews.
As Pierre and I validated the refacto and tests, I'm merging it.

File: datamodel.customer-survey.xml
Comment (jbostoen) [inline datamodel.customer-survey.xml:1970]: We may need to revisit this. Is the $ sign enough here?
I think sometimes the encoded version (%24 instead of $) is passed. So we'd need to adjust this regex.

Comment (Molkobain) [reply datamodel.customer-survey.xml:1970]: Indeed, you can find a similar behavior in `\MetaModel::ApplyParams()`

**Pull request merged**
